tso-server/src/main/java/org/apache/omid/tso/ReplyProcessorImpl.java

```diff
@@ -180,9 +192,19 @@ public void manageResponsesBatch(long batchSequence, Batch batch) {
 
     }
 
-    @Override
-    public void sendCommitResponse(long startTimestamp, long commitTimestamp, Channel c, MonitoringContext monCtx) {
+    @VisibleForTesting
+    void updateLowWatermark(Optional<Long> newLowwatermark) {
+        //TODO YONIGO - validate this is single threaded
+        if (newLowwatermark.isPresent() && newLowwatermark.get() > highestLowWaterMarkSeen) {
+            highestLowWaterMarkSeen = newLowwatermark.get();
+            lowWatermarkWriter.persistLowWatermark(highestLowWaterMarkSeen);
```

Comment:
Basically, you can write the low water mark only once per batch, just collect the highest as you do in here and write it, in case it exists, once the batch terminates.